Pack.targets fix

jp2masa opened the pull request:
## Bug
Fixes: NuGet/Home#6099
Regression: Probably not
If Regression then when did it last work:  N/A
If Regression then how are we preventing it in future:  N/A

## Fix
Details: The nuspec path for incremental build was wrong. Fixed it.

## Testing/Validation
Tests Added: No  
Reason for not adding tests: Not sure if targets should have tests
Validation done: None

dnfclas commented:
@jp2masa,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

jainaashish commented:
Thanks for the PR, can you also create a issue here https://github.com/NuGet/home/issues to track this change? And also add some test for it.

jp2masa commented:
I created the issue to track the change, but I don't know how to test it and I think that currently there are no tests for Pack.targets.

rohit21agrawal commented:
The tests are in dotnet.integration.test . The change looks good

jp2masa commented:
I tried adding a test, but the tests use an old version of Pack.targets (older than the one in the `dev` branch).

rohit21agrawal commented:
Make sure you build from the command line first and then run test

jp2masa commented:
Added the test and fixed one more issue, if the paths were set as global properties, they could be missing a trailing slash, which would break NuGet Pack outputs.

rohit21agrawal reviewed (changes requested):
i just realized this approach won't work since the package version gets normalized for the output file name.

use this existing target instead to get the list of items that the pack target produces:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L63

rohit21agrawal commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -169,19 +169,12 @@ Copyright (c) .NET Foundation. All rights reserved.
-                             OR '$(IncludeSymbols)' == 'true'"/>
     </ItemGroup>
+
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
```
@jp2masa sorry for getting back to you on this so late (i was on vacation). we avoid invoking msbuild in targets where it can be avoided as it might break incremental builds.

what i would suggest is to make the target _CalculateInputsOutputsForPack depend on the target _GetOutputItemsFromPack which will make @(_OutputPackItems) available to you in the _CalculateInputsOutputsForPack target

You can then just write NuGetPackOutput Include="@(_OutputPackItems)

rohit21agrawal commented:
@jp2masa just ran this against the CI, and your test fails:

```at Dotnet.Integration.Test.PackCommandTests.PackCommand_PackTarget_IsIncremental(String tfmProperty, String tfmValue) in E:\A\_work\r1\a\NuGet.Client\test\NuGet.Core.FuncTests\Dotnet.Integration.Test\PackCommandTests.cs:line 1981```

Assert.Equal() Failure
Expected: 2017-11-22T04:10:55.5758705Z
Actual: 2017-11-21T20:10:55.5758705-08:00

jp2masa commented:
I'll fix that ASAP.
Is there any difference between comparing UTC times or non UTC times?

rohit21agrawal commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -1940,5 +1940,49 @@ public void PackCommand_PackProject_OutputsBuildActionForContentFiles(string ite
+                Assert.True(File.Exists(nupkgPath), "The output .nupkg is not in the expected place");
+                Assert.True(File.Exists(nuspecPath), "The intermediate nuspec file is not in the expected place");
+
+                Assert.Equal(nupkgLastWriteTime, File.GetLastWriteTime(nupkgPath));
```
you just need to replace this with File.GetLastWriteTimeUtc like you have done above

jp2masa committed:
Pack fixes.

rohit21agrawal commented:
@jp2masa appreciate your contribution to make NuGet better!
![](https://thumbs.gfycat.com/FavoriteOrangeBeaver-size_restricted.gif)

rohit21agrawal commented:
https://thumbs.gfycat.com/FavoriteOrangeBeaver-size_restricted.gif
